amm/contracts/stable_pool/lib.rs

```diff
@@ -461,9 +502,49 @@ pub mod stable_pool {
             // mint admin fee
             self.mint_admin_fee(fee, token_out_id)?;
 
-            // note that fee is applied to token_out (same as in _swap_to)
+            // transfer token_in
+            _ = self._transfer_in(token_in_id, Some(token_in_amount))?;
+
+            // transfer token_out
+            self.token_by_address(token_out)
+                .transfer(to, token_out_amount, vec![])?;
+
+            self.env().emit_event(Swap {
+                sender: self.env().caller(),
+                token_in,
+                amount_in: token_in_amount,
+                token_out,
+                amount_out: token_out_amount,
+                to,
+            });
+            self.env().emit_event(Sync {
+                reserves: self.reserves(),
+            });
+            // note that fee is applied to token_out (same as in _swap_exact_in)
             Ok((token_in_amount, fee))
         }
+
+        fn _transfer_in(
+            &self,
+            token_id: usize,
+            amount: Option<u128>,
+        ) -> Result<u128, StablePoolError> {
+            let mut token = self.token_by_id(token_id);
+            if let Some(token_amount) = amount {
```

Comment:
if you threw an error here on `token_amount == 0` then you could remove it from the other places